lib/TPP/Dialect/Tpp/TppOps.cpp

```diff
@@ -17,14 +17,167 @@
 using namespace mlir;
 using namespace mlir::tpp;
 
+//===----------------------------------------------------------------------===//
+// Utils
+//===----------------------------------------------------------------------===//
+
+static void printCommaSeparatedList(OpAsmPrinter &printer, ValueRange values) {
+  printer << '(';
+  for (auto [idx, value] : llvm::enumerate(values)) {
+    printer << value << " : " << value.getType();
+    if (idx != values.size() - 1)
+      printer << ", ";
+  }
+  printer << ')';
+}
+
+static ParseResult parseTppOp(OpAsmParser &parser, OperationState &result) {
+  SmallVector<OpAsmParser::UnresolvedOperand> operands;
+  SmallVector<Type> operandsTypes;
+
+  bool isMemRef = false;
+  if (succeeded(parser.parseOptionalKeyword("ins")))
+    isMemRef = true;
+
+  // Parse operands.
+  SmallVector<llvm::SMLoc> locsOperands;
+  auto parseOperand = [&]() -> ParseResult {
+    locsOperands.push_back(parser.getCurrentLocation());
+    if (parser.parseOperand(operands.emplace_back()) ||
+        parser.parseColonType(operandsTypes.emplace_back()))
+      return failure();
+    return success();
+  };
+
+  if (parser.parseCommaSeparatedList(OpAsmParser::Delimiter::Paren,
+                                     parseOperand)) {
+    return failure();
+  }
+
+  if (isMemRef) {
+    if (parser.parseKeyword("outs") || parser.parseLParen() ||
+        parser.parseOperand(operands.emplace_back()) ||
+        parser.parseColonType(operandsTypes.emplace_back()) ||
+        parser.parseRParen())
+      return failure();
+  } else {
+    // Parse result types.
+    SmallVector<Type> resultTypes;
+    llvm::SMLoc resultTypeLoc = parser.getCurrentLocation();
+    if (parser.parseArrowTypeList(resultTypes) ||
+        parser.addTypesToList(resultTypes, result.types))
+      return failure();
+
+    if (resultTypes.size() != 1) {
+      return parser.emitError(resultTypeLoc,
+                              "expect single result at tensor abstraction");
+    }
+  }
+
+  // Validate operands. Scan each operand one-by-one to emit
+  // better diagnostic.
+  for (auto [idx, operand] : llvm::enumerate(operands)) {
+    if (parser.resolveOperand(operand, operandsTypes[idx], result.operands))
+      return failure();
+    if (isMemRef && operandsTypes[idx].isa<RankedTensorType>())
+      return parser.emitError(locsOperands[idx], "expect memref type");
+    if (!isMemRef && operandsTypes[idx].isa<MemRefType>())
+      return parser.emitError(locsOperands[idx], "expect tensor type");
+  }
+
+  NamedAttrList attrs;
+  if (parser.parseOptionalAttrDictWithKeyword(attrs))
+    return failure();
+  result.addAttributes(attrs);
+  return success();
+}
+
+// Print a tpp op. Note that `out` can be null. It is null for unary and binary
+// at tensor abstraction. Ternary operations have `out` also at tensor
+// abstraction, that is the init value (ternary ops are +=).
```

Comment:
I think this comment is a bit outdated.

Reply:
why? which part are you referring to?

Reply:
There are no `outs` at tensor level but now I guess you refer to this abstract concept that like `C` in matmul is an "out". It can get a bit confusing at times 😅